### PR TITLE
restore: auto tune concurrency configuration when using new mode restoration

### PR DESCRIPTION
**Pull request by 3pointer**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/50701
Problem Summary:
make concurrency easy to use for user.

### What changed and how does it work?
based on tikv config `import.num-threads` to determine the concurrency.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/conn/conn.go [hunk 1]
```diff
@@ -287,42 +293,56 @@ func (mgr *Mgr) GetTS(ctx context.Context) (uint64, error) {
...
-		rs, e := units.RAMInBytes(c.Cop.RegionSplitSize)
-		if e != nil {
-			return e
+		if !mergeRegionSize.HasSet || !mergeRegionKeyCount.HasSet {
+			size, keys, e := kvconfig.ParseMergeRegionSizeFromConfig(respBytes)
+			if e != nil {
+				log.Warn("Failed to parse region split size and keys from config", logutil.ShortError(e))
+				return e
+			}
+			if mergeRegionKeyCount.Value == DefaultMergeRegionKeyCount || keys < mergeRegionKeyCount.Value {
```
Comment (Leavrth) [inline br/pkg/conn/conn.go:320]: why choose the `region split key` instead of `region split size`?

File: br/pkg/conn/conn.go [hunk 2]
```diff
@@ -287,42 +293,56 @@ func (mgr *Mgr) GetTS(ctx context.Context) (uint64, error) {
...
-			regionSplitSize = urs
-			regionSplitKeys = c.Cop.RegionSplitKeys
+		if !importGoroutines.HasSet {
+			threads, e := kvconfig.ParseImportThreadsFromConfig(respBytes)
+			if e != nil {
+				log.Warn("Failed to parse import num-threads from config", logutil.ShortError(e))
+				return e
+			}
+			// We use 8 times the default value because it's an IO-bound case.
+			if importGoroutines.Value == DefaultImportNumGoroutines || (threads > 0 && threads*8 < importGoroutines.Value) {
```
Comment (Leavrth) [inline br/pkg/conn/conn.go:332]: Does `threads > 0 && threads*8 < importGoroutines.Value` means the maximum of `importGoroutines.Value` is `DefaultImportNumGoroutines`?
It seems not large enough.

Comment (3pointer) [reply br/pkg/conn/conn.go:332]: No, when first time reach here. the `importGoroutines.Value` must be updated.

File: br/pkg/conn/conn.go (hunk 1 above)
Comment (3pointer) [reply br/pkg/conn/conn.go:320]: becase I think the both values are updated together.

**Review by Leavrth — APPROVED** [review]

File: br/pkg/config/kv.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
+// Copyright 2024 PingCAP, Inc. Licensed under Apache-2.0.
+package config
+
+import (
+	"encoding/json"
+
+	"github.com/docker/go-units"
+)
+
+type ConfigSet[T uint | uint64] struct {
```
Comment (YuJuncen) [inline br/pkg/config/kv.go:10]: `ConfigSet` sounds like a set of config values. Maybe `ConfigTerm`.

File: br/pkg/config/kv.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	Value  T
+	HasSet bool
```
Comment (YuJuncen) [inline br/pkg/config/kv.go:12]: Maybe `HasBeenSet` or `Modified`.

File: br/pkg/config/kv.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	var c config
+	e := json.Unmarshal(resp, &c)
+	if e != nil {
+		return 0, e
+	}
+
+	return c.Import.Threads, nil
+}
+
+func ParseMergeRegionSizeFromConfig(resp []byte) (uint64, uint64, error) {
```
Comment (YuJuncen) [inline br/pkg/config/kv.go:38]: It seems this function is only called in `conn` package and doesn't rely on any item in `config`. Why not make it a private function in `conn`?

File: br/pkg/task/restore_raw.go [hunk 6]
```diff
@@ -7,6 +7,7 @@ import (
 
 	"github.com/pingcap/errors"
 	"github.com/pingcap/log"
+	kvconfig "github.com/pingcap/tidb/br/pkg/config"
```
Comment (YuJuncen) [inline br/pkg/task/restore_raw.go:10]: Why rename it to `kvconfig`? It seems belongs to BR. Perhaps use `pconfig` as `restore.go` did.

File: br/pkg/task/restore.go [hunk 7]
```diff
@@ -121,17 +121,17 @@ type RestoreCommonConfig struct {
...
-		cfg.MergeSmallRegionSizeBytes = conn.DefaultMergeRegionSizeBytes
+	if !cfg.MergeSmallRegionSizeBytes.HasSet {
+		cfg.MergeSmallRegionSizeBytes.Value = conn.DefaultMergeRegionSizeBytes
 	}
 	if len(cfg.Granularity) == 0 {
 		cfg.Granularity = string(restore.FineGrained)
 	}
-	if cfg.ConcurrencyPerStore == 0 {
-		cfg.ConcurrencyPerStore = 128
+	if cfg.ConcurrencyPerStore.HasSet {
```
Comment (YuJuncen) [inline br/pkg/task/restore.go:133]: It is strange to reset to default when the value has been set? Just typo?
```suggestion
	if !cfg.ConcurrencyPerStore.HasSet {
```

File: br/pkg/config/kv.go (hunk 5 above)
Comment (3pointer) [reply br/pkg/config/kv.go:38]: I'd like to put all parse config logic into one pkg for future usage.

File: br/pkg/task/restore_raw.go (hunk 6 above)
Comment (3pointer) [reply br/pkg/task/restore_raw.go:10]: I think all logic belong to TiKV could use this name

File: br/pkg/config/kv.go (hunk 3 above)
Comment (3pointer) [reply br/pkg/config/kv.go]: done

**Review by YuJuncen — APPROVED** [review]

**Pull request merged**